cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/ExecutableBuilder.java

```diff
@@ -244,39 +242,34 @@ public ExecutableModellingResult transformToExecutable(ParsedSlang parsedSlang,
                                           String namespace, String execName) {
+
+        Map<String, Map<String, Object>> onFailureStepData = preCompileValidator.validateOnFailurePosition(
+                workFlowRawData,
+                execName,
+                errors
+        );
+
         Workflow onFailureWorkFlow = null;
-        List<Map<String, Map<String, Object>>> onFailureData;
-        Iterator<Map<String, Map<String, Object>>> stepsIterator = workFlowRawData.iterator();
-        boolean onFailureFound = false;
-        while (stepsIterator.hasNext()) {
-            Map<String, Map<String, Object>> stepData = stepsIterator.next();
-            String stepName = stepData.keySet().iterator().next();
-            if (stepName.equals(ON_FAILURE_KEY)) {
-                if (onFailureFound) {
-                    errors.add(new RuntimeException("Flow: '" + execName + "' syntax is illegal.\n" + MULTIPLE_ON_FAILURE_MESSAGE_SUFFIX));
-                } else {
-                    onFailureFound = true;
-                }
-                try {
-                    //noinspection unchecked
-                    onFailureData = (List<Map<String, Map<String, Object>>>) stepData.values().iterator().next();
-                } catch (ClassCastException ex) {
-                    onFailureData = new ArrayList<>();
-                    errors.add(new RuntimeException("Flow: '" + execName + "' syntax is illegal.\nBelow 'on_failure' property there should be a list of steps and not a map"));
+        if (MapUtils.isNotEmpty(onFailureStepData)) {
+            List<Map<String, Map<String, Object>>> onFailureData;
+            try {
+                //noinspection unchecked
+                onFailureData = (List<Map<String, Map<String, Object>>>) onFailureStepData.values().iterator().next();
+            } catch (ClassCastException ex) {
+                onFailureData = new ArrayList<>();
+                errors.add(new RuntimeException("Flow: '" + execName + "' syntax is illegal.\nBelow 'on_failure' property there should be a list of steps and not a map"));
+            }
+            if (CollectionUtils.isNotEmpty(onFailureData)) {
+                if (onFailureData.size() > 1) {
+                    errors.add(new RuntimeException("Flow: '" + execName + "' syntax is illegal.\nBelow 'on_failure' property there should be only one step"));
                 }
-                if (CollectionUtils.isNotEmpty(onFailureData)) {
-                    if (onFailureData.size() > 1) {
-                        errors.add(new RuntimeException("Flow: '" + execName + "' syntax is illegal.\nBelow 'on_failure' property there should be only one step"));
-                    }
-                    handleOnFailureStepNavigationSection(onFailureData, execName, errors);
+                handleOnFailureStepNavigationSection(onFailureData, execName, errors);
```

Comment:
I intended to to validate in `validateOnFailurePosition` only data related to the position of the step: multiple appearance and last step in the flow -- then at a later stage we can do the rest of the validations regarding structure

so, I am leaving it as is for now - thanks :+1: